<launch>
  <!-- Launch Parameters -->
  <arg name="cameraName" default="camera" />
  <arg name="targetName" default="ugv0" />
  <arg name="recordData" default="false" /> <!-- record to rosbag -->
  <arg name="ekf" default="false" />
  <arg name="estimator" default="true" />
  <arg name="normalizedKinematics" default="false" />
  <arg name="predictor" default="true" /> <!-- use predictor instead of zero order hold in estimation algorithm -->
  <arg name="deadReckoning" default="true" /> <!-- use turtlebot onboard odometry for rotation estimation -->
  <arg name="artificialSwitching" default="false" /> <!-- artificial switching (i.e. fake loss of visibility) -->
  <arg name="visibilityTimeout" default="0.1" /> <!-- time to wait before instantiating predictor or zero order hold -->
  <arg name="delTon" default="4.0" /> <!-- contiguous visibility time if artificial switching -->
  <arg name="delToff" default="1.0" /> <!-- contiguous invisibility time if artificial switching -->
  <arg name="X0" default="2.0" /> <!-- Initial condition -->
  <arg name="Y0" default="-1.0" /> <!-- Initial condition -->
  <arg name="Z0" default="10.0" /> <!-- Initial condition -->
  <arg name="useVelocityMap" default="true" /> <!-- use velocity map, or communicate target velocities -->
  
  <!-- Estimator -->
  <node if="$(arg estimator)" name="exp" pkg="switch_vis_exp" type="exp_node" output="screen" >
    <param name="cameraName" value="$(arg cameraName)" />
    <param name="targetName" value="$(arg targetName)" />
    <param name="usePredictor" value="$(arg predictor)" />
    <param name="deadReckoning" value="$(arg deadReckoning)" />
    <param name="artificialSwitching" value="$(arg artificialSwitching)" />
    <param name="useVelocityMap" value="$(arg useVelocityMap)" />
    <param name="visibilityTimeout" value="$(arg visibilityTimeout)" />
    <param name="k1" value="2.0" /> <!-- Gains -->
    <param name="k2" value="2.0" /> <!-- Gains -->
    <param name="k3" value="1.0" /> <!-- Gains -->
    <param name="delTon" value="$(arg delTon)" />
    <param name="delToff" value="$(arg delToff)" />
    <param name="X0" value="$(arg X0)" />
    <param name="Y0" value="$(arg Y0)" />
    <param name="Z0" value="$(arg Z0)" />
    <remap from="output" to="exp/output" />
    <remap from="output_point" to="exp/output_point" />
  </node>
  
  <!-- EKF -->
  <node if="$(arg ekf)" name="ekf" pkg="switch_vis_exp" type="ekf_node" output="screen" >
    <param name="cameraName" value="$(arg cameraName)" />
    <param name="targetName" value="$(arg targetName)" />
    <param name="deadReckoning" value="$(arg deadReckoning)" />
    <param name="normalizedKinematics" value="$(arg normalizedKinematics)" />
    <param name="artificialSwitching" value="$(arg artificialSwitching)" />
    <param name="useVelocityMap" value="$(arg useVelocityMap)" />
    <param name="visibilityTimeout" value="$(arg visibilityTimeout)" />
    <param name="q" value="0.5" />
    <param name="r" value="1.0" />
    <param name="delTon" value="$(arg delTon)" />
    <param name="delToff" value="$(arg delToff)" />
    <param name="X0" value="$(arg X0)" />
    <param name="Y0" value="$(arg Y0)" />
    <param name="Z0" value="$(arg Z0)" />
    <remap from="output" to="ekf/output" />
    <remap from="output_point" to="ekf/output_point" />
  </node>
  
  <!-- Record data -->
  <node if="$(arg recordData)" name="record" pkg="rosbag" type="record" args="-o /home/ncr/experiment_data/predictor/exp_Pred_velMap_random.bag /ekf/output /exp/output /markerImage /$(arg cameraName)/camera_info /markerCenters /ugv0/pose /ugv0/odom /image/pose /image/body_vel /tf" />
</launch>
